src/objective-c/GRPCClient/private/GRPCCore/GRPCChannel.m

```diff
@@ -50,32 +51,48 @@ - (instancetype)initWithHost:(NSString *)host callOptions:(GRPCCallOptions *)cal
 }
 
 - (id<GRPCChannelFactory>)channelFactory {
-  GRPCTransportType type = _callOptions.transportType;
-  switch (type) {
-    case GRPCTransportTypeChttp2BoringSSL:
-      // TODO (mxyan): Remove when the API is deprecated
-#ifdef GRPC_COMPILE_WITH_CRONET
-      if (![GRPCCall isUsingCronet]) {
-#else
-    {
-#endif
-        NSError *error;
-        id<GRPCChannelFactory> factory = [GRPCSecureChannelFactory
-            factoryWithPEMRootCertificates:_callOptions.PEMRootCertificates
-                                privateKey:_callOptions.PEMPrivateKey
-                                 certChain:_callOptions.PEMCertificateChain
-                                     error:&error];
-        NSAssert(factory != nil, @"Failed to create secure channel factory");
-        if (factory == nil) {
-          NSLog(@"Error creating secure channel factory: %@", error);
+  if (_callOptions.transport != NULL) {
+    id<GRPCTransportFactory> transportFactory =
+        [[GRPCTransportRegistry sharedInstance] getTransportFactoryWithId:_callOptions.transport];
+    if (!
+        [transportFactory respondsToSelector:@selector(createCoreChannelFactoryWithCallOptions:)]) {
+      // impossible because we are using GRPCCore now
+      [NSException raise:NSInternalInconsistencyException
+                  format:@"Transport factory type is wrong"];
+    }
+    id<GRPCCoreTransportFactory> coreTransportFactory =
+        (id<GRPCCoreTransportFactory>)transportFactory;
+    return [coreTransportFactory createCoreChannelFactoryWithCallOptions:_callOptions];
+  } else {
+    // To maintain backwards compatibility with tranportType
+    GRPCTransportType type = _callOptions.transportType;
+    switch (type) {
+      case GRPCTransportTypeChttp2BoringSSL:
+        // TODO (mxyan): Remove when the API is deprecated
+        {
+          NSError *error;
+          id<GRPCChannelFactory> factory = [GRPCSecureChannelFactory
+              factoryWithPEMRootCertificates:_callOptions.PEMRootCertificates
+                                  privateKey:_callOptions.PEMPrivateKey
+                                   certChain:_callOptions.PEMCertificateChain
+                                       error:&error];
+          NSAssert(factory != nil, @"Failed to create secure channel factory");
+          if (factory == nil) {
+            NSLog(@"Error creating secure channel factory: %@", error);
+          }
+          return factory;
         }
-        return factory;
+      case GRPCTransportTypeCronet: {
+        id<GRPCCoreTransportFactory> transportFactory = (id<GRPCCoreTransportFactory>)[
```

Comment:
Because secure and insecure transports are the default ones when Core is included.

If this method is executed, `GRPCChannel.m` exists in the process, hence secure and insecure transports are for sure registered. But Cronet might not be included as a dependency of this process. So we have to query the registry in this case.